Source/DafnyCore/AST/DafnyAst.cs

```diff
@@ -77,108 +78,114 @@ public abstract class INode {
+              if (node is null) {
+                return;
+              }
+
+              if (node.tok.Filename != tok.Filename || node is Expression { IsImplicit: true } ||
+                  node is DefaultValueExpression) {
+                // Ignore any auto-generated expressions.
+              } else if (node != this && node.RangeToken != null) {
+                UpdateStartEndToken(node.StartToken);
+                UpdateStartEndToken(node.EndToken);
+              } else {
+                UpdateStartEndToken(node.tok);
+              }
+
+              node.Children.Iter(updateStartEndTokRecursive);
             }
-          }
 
-          void UpdateStartEndTokRecursive(INode node) {
-            if (node is null) {
-              return;
-            }
+            updateStartEndTokRecursive(this);
 
-            if (node.tok.Filename != tok.Filename || node is Expression { IsImplicit: true } ||
-                node is DefaultValueExpression) {
-              // Ignore any auto-generated expressions.
-            } else if (node != this && node.GetRangeToken() != null) {
-              UpdateStartEndToken(node.GetStartToken());
-              UpdateStartEndToken(node.GetEndToken());
-            } else {
-              UpdateStartEndToken(node.tok);
+            if (FormatTokens != null) {
+              foreach (var token in FormatTokens) {
+                UpdateStartEndToken(token);
+              }
             }
 
-            node.Children.Iter(UpdateStartEndTokRecursive);
-          }
-
-          UpdateStartEndTokRecursive(this);
-
-          if (FormatTokens != null) {
-            foreach (var token in FormatTokens) {
-              UpdateStartEndToken(token);
-            }
+            rangeToken = new RangeToken(startTok, endTok);
           }
+        }
 
-          rangeToken = new RangeToken(startTok, endTok);
+        if (rangeToken.Filename == null) {
+          rangeToken.Filename = tok.Filename;
         }
-      }
 
-      if (rangeToken.Filename == null) {
-        rangeToken.Filename = tok.Filename;
+        return rangeToken;
       }
-
-      return rangeToken;
+      set => rangeToken = value;
     }
 
-    public IToken GetStartToken() => GetRangeToken()?.StartToken;
-    public IToken GetEndToken() => GetRangeToken()?.EndToken;
+    [DebuggerBrowsable(DebuggerBrowsableState.Never)]
+    public IToken StartToken => RangeToken?.StartToken;
+
+    [DebuggerBrowsable(DebuggerBrowsableState.Never)]
+    public IToken EndToken => RangeToken?.EndToken;
 
     protected IReadOnlyList<IToken> OwnedTokensCache;
 
     /// <summary>
     /// A token is owned by a node if it was used to parse this node,
     /// but is not owned by any of this Node's children
     /// </summary>
-    public IEnumerable<IToken> GetOwnedTokens() {
-      if (OwnedTokensCache != null) {
-        return OwnedTokensCache;
-      }
+    [DebuggerBrowsable(DebuggerBrowsableState.Never)]
+    public IEnumerable<IToken> OwnedTokens {
```

Comment:
I don't think `OwnedTokens` should be a property since it does a computation. I would like to follow the MS guidelines that I linked to before, and not have any properties that are orders of magnitude slower than a field access.

Reply:
`OwnedTokens` is as fast as a field access after the first time, and the code called only once does not rely on I/O or asynchronous programming, and its total execution (over all the nodes) is linear in terms of the length of the code.

I sometimes call them "lazy properties" because I'm used in Scala and in Scala this is a standard practice.
In C#, we call such properties "Cached Properties", and it's acceptable that they are doing computation the first time, e.g. I don't see a comment in any of the answers and threads there that complains that any of these properties should be a method instead of a property because it does computation the first time https://stackoverflow.com/questions/2579329/cached-property-easier-way

Note that we could make it even more concise using techniques such as described here
https://stackoverflow.com/a/63085705/1287856

Reply:
> I don't see a comment in any of the answers and threads there that complains that any of these properties should be a method instead of a property because it does computation the first time

This one does: https://stackoverflow.com/a/2579369/93197

I'm OK with having cached computation properties. I think it's safer to use methods for anything that might trigger a computation, so the programmer is made aware of that, but I guess if you use a cached property in a hot loop, the caching will limit the amount of computation done and the hotness of the loop is irrelevant.

Reply:
> This one does: https://stackoverflow.com/a/2579369/93197

Good finding. It's a least upvoted answer, but it's an answer.

> Who's we?

I and those C# developers who are not bothered by a property having an extra computation the first time :-)

> I think we should use C# source generators ([link](https://learn.microsoft.com/en-us/dotnet/csharp/roslyn-sdk/source-generators-overview)) to write a [Cached] attribute that you can slap on any property (or method even), to make it cached.

We can do that if we really feel it's needed. Or use the more idiomatic Lazy<>